File: oop-exercises/encapsulation/src/service/StudentService.ts
import { Inject, Injectable } from '@nestjs/common';
import { StudentRepository } from '../repository/StudentRepository';
import { Student } from '../entity/Student';

@Injectable()
export class StudentService {
  constructor(@Inject('StudentRepository') private readonly studentRepository: StudentRepository) {}

  async create(id: number, name: string, score: number): Promise<void> {
    const student = new Student();
    student.id = id;
    student.name = name;
    student.score = score;

    if (score < 0 || score > 100) {
      throw new Error('Score must be between 0 and 100');
    }

    if (score > 95) {
      student.isEligibleForFreeTuition = true;
    } else {
      student.isEligibleForFreeTuition = false;
    }

    await this.studentRepository.save(student);
  }
}

File: oop-encapsulation/ruby/src/service/student_service.rb

require_relative '../repository/student_repository'
require_relative '../entity/student'

class StudentService
  def initialize(studentRepository)
    @studentRepository = studentRepository
  end

  def create(id, name, score)
    student = Student.new
    student.id = id
    student.name = name
    student.score = score

    if score < 0 || score > 100
      raise 'Score must be between 0 and 100'
    end

    if score > 95
      student.is_eligible_for_free_tuition = true
    else
      student.is_eligible_for_free_tuition = false
    end

    @studentRepository.save(student)
  end
end
